CefSharp.Core/Internals/ClientAdapter.cpp

```diff
@@ -323,15 +350,36 @@ namespace CefSharp
 
         bool ClientAdapter::OnBeforeBrowse(CefRefPtr<CefBrowser> browser, CefRefPtr<CefFrame> frame, CefRefPtr<CefRequest> request, bool isRedirect)
         {
-            IRequestHandler^ handler = _browserControl->RequestHandler;
-            if (handler == nullptr)
+            if (browser->IsPopup())
             {
-                return false;
+                auto popupHandler = _browserControl->PopupHandler;
+                if (popupHandler != nullptr)
+                {
+                    auto browserId = browser->GetIdentifier();
+                    IBrowser^ entry;
+                    if (_popupBrowsers->TryGetValue(browserId, entry))
+                    {
+                        auto frameWrapper = gcnew CefFrameWrapper(frame, _browserAdapter);
+                        CefRequestWrapper^ wrapper = gcnew CefRequestWrapper(request);
+                        return popupHandler->OnBeforeBrowse(_browserControl, entry, wrapper, isRedirect, gcnew CefFrameWrapper(frame, _browserAdapter));
```

Comment:
The overhead of `CefFrameWrapper` is obviously very small.  The easiest thing for us to do is to dispose of the wrappers after we finish the callback. 

Do we get notifications when frames/browsers get destroyed? If we did, we could do something more complicated:

* Have an internal .Net friendly `CefFrameWrapper` that we cache, and is fronted by a weak reference owning `ClientCefFrameWrapper` like object.
i.e. `ClientCefFrameWrapepr` would contain a GC weak reference to our `CefFrameWrapper` instance. 

This way, we can own the lifetimes of the important CEF ref counts via our `CefFrameWrapper`, and if our users hang around to the `ClientCefFrameWrapper`s too long, then they'll get the appropriate exception afterwards when the weak reference is no longer valid.

It is even spookier action at a distance, but at least it lets us think more clearly about how we're managing CEF object lifetimes. (which clearly has given us trouble in the past, i.e. shutdowns)

Thoughts?

I'm not sure `OnLoadingStateChange` is the correct context for disposing of the instances. (That would surprise me, since we don't cache any information in our .Net wrappers, we always ask CEF what the current values are.)

Bill